### ISSUE TITLE
Using ssh instead of wslbridge

### ISSUE THREAD

**Issue by AsharLohmar** [mintty/wsltty#99]
Seeing that with the latest version we can have sshd inside WSL running as a service (without no window opened), I was wondering if it would be a good idea to just ssh to the WSL instead of using wslbridge.
```
#start sshd
ubuntu.exe run sudo service ssh start
# mintty+ssh
mintty.exe -h error /bin/ssh ashar@127.0.0.1
```

I've got the necessary files from a cygwin install, but maybe the ssh could also be built without shared libs.
```
ssh.exe
cygcom_err-2.dll
cygcrypto-1.0.0.dll
cyggcc_s-seh-1.dll
cyggssapi_krb5-2.dll
cygiconv-2.dll
cygintl-8.dll
cygk5crypto-3.dll
cygkrb5-3.dll
cygkrb5support-0.dll
cygz.dll
```

**Comment by mintty** [comment]
And it would depend on prior setup of sshd, so it would not offer seamless user experience.
Still an interesting option, however, if static linking of ssh works. Maybe I'll add it to the wiki.

**Comment by AsharLohmar** [comment]
@Biswa96 seeing that/how Microsoft "insists" on pushing the updates that should not be a problem (for long).
I know that this would require (a lot) more work in order to setup everything. It would require sshd running on different ports if more distros are installed. 
Seemed like a more "natural" way to work, but I was just wondering if it could be a "feasible" option ... would it have any benefit or downside on the performance or the functionality support.

**Comment by AsharLohmar** [comment]
Yes, I've got the connection working.
In the first message I've put the minimum required "stuff" to get the things "rolling".
I've also:
 - modified the sshd_config inside WSL to put ssh in listen only on 127.0.0.1 and to enable/allow password auth
 - created a home folder for my user  (/home/<user> at the same level as the /bin folder) to get rid of some warning messages.

I've also got key auth working
 - I already had a key in my WSL, I've used ssh-copy-id to get it into the authorized_keys, so I've copied that key inside the .ssh folder in the home folder I've created before
(at this point I was getting warnings about the file permissions being too "open")
 - I've took the chmod from a cygwin install and using dash as a shell I've changed the permissions of the key to 600

Being weekend and all, I haven't extensively used it in this way yet, but I will in the following days. I'll let you know if I encounter any issues.

**Comment by AsharLohmar** [comment]
As said before, I needed the chmod in order to set the right permission on the key file for the key auth (maybe there is some other way to do it), otherwise that's the minimum setup.

I guess that maybe `ssh-keygen` and `ssh-copy-id` could be useful in the provisioning fase (first login), to create a (new) key locally in the wsltty home and transfer it (the public part) inside WSL. Might even fix the key permissions issue, by creating it with the right permissions from the beginning.

**Comment by AsharLohmar** [comment]
Found an "issue", I'll keep adding them here to keep the thread clean :
- windows `PATH` doesn't get merged inside the wsl `PATH`.

**Comment by Biswa96** [comment]
To use `ssh` with WSL users have to configure the ssh server in WSL. The steps:

* WSL:
  - Remove openssh-server: `apt remove openssh-server`.
  - Install openssh-server: `apt install openssh-server`.
  - File edited: `/etc/ssh/sshd_config`.

* Windows(cygwin):
  - Inbound Firewall rule for `\rootfs\usr\sbin\sshd`.
  - cygwin command: `ssh user@host -p port`.
  - File saved: `cygwin/home/user/.ssh/known-hosts`.

But with wslbridge users don't have to do that. wslbridge itself opens sockets in localhost and connect both backend and frontend. Like @rprichard said:
>"wslbridge is a Cygwin program that allows connecting to the WSL command-line environment over TCP sockets, as with ssh, but without the overhead of configuring an SSH server."

**Comment by Biswa96** [comment]
>"However, if static linking of ssh works."

@mintty `ssh` can be statically linked but the size of that executable becomes 2.5 MB approximately. I follow this https://github.com/shadowsocks/shadowsocks-libev/issues/623#issuecomment-216627477

First download x64 openssh source from https://mirrors.kernel.org/sourceware/cygwin/x86_64/release/openssh/ (or any other mirror). Use this command with `./configure` shell script:

    LIBS="-ldl -lz" LDFLAGS="-Wl,-static -static -static-libgcc -s" ./configure --prefix=/opt

Then use `make` command to compile. Only `cygwin1.dll` remains.

**Comment by mintty** [comment]
As discussed, offering an option to use ssh instead of wslbridge within the wsltty setup would not have much benefit but a lot of potential trouble, so I don't plan that.

On the other hand, mintty can be configured to run ssh to connect to WSL already, but would not apply various WSL settings in that case (like rootfs for path name conversions, icon selection).
It may be useful to have a version of the WSL option that figures out those implicit options and applies them, while child invocation is not affected (so you can use what you like).

**Comment by Biswa96** [comment]
Ya, agree with you. First users have to configure the ssh server in WSL side. Then connect mintty with ssh. But with wslbrisdge users don't have to do that, just run a command.

**Comment by Biswa96** [comment]
Recently, I'm exposing the COM interface of WSL with which mintty may not require wslbridge/ssh/pty stuffs. See [WSL_Reverse](https://github.com/Biswa96/WSL_Reverse).

**Comment by evaryont** [comment]
Does it change things now that OpenSSH is shipped as part of Windows? `C:\Windows\System32\OpenSSH\ssh.exe` is included as of the April 2018 update, along with curl (`C:\Windows\System32\curl.exe`).

**Comment by mintty** [comment]
New optoin `--WSLmode[=distro]` sets up mintty behaviour to support the given (or default) WSL distribution but (unlike option `--WSL`) it does not actually launch WSL. So you have an opportunity to set up your own launching method, using your preferred launching gateway.

**Comment by mintty** [comment]
@evaryont, using a Windows console program as gateway would not be sufficient as it is part of the task of the gateway to connect to the pty-based terminal frontend. In this case another frontend would be needed, e.g. wslbridge :) Whether using a dual gateway achieves any benefit as compared to the current wslbridge-only solution, I have no idea.

**Comment by mintty** [comment]
Concerning `--WSLmode`, mintty could provide the determined WSL GUID and perhaps the file system root in environment variables to the launcher. If that's desired for any such solution, drop a request here.

**Comment by Biswa96** [comment]
What are the values of those long options on option struct? GitHub shows it as a black box.

**Comment by mintty** [comment]
You mean values for `--WSLmode`? Same as for `--WSL`; I didn't mention any and I don't see a black box.

**Comment by Biswa96** [comment]
Here the section from `winmain.c`:
```c
#if CYGWIN_VERSION_API_MINOR >= 74
  {"WSL",        optional_argument, 0, ''},  // short option not enabled
  {"WSLmode",    optional_argument, 0, ''},  // short option not enabled
```
Here you can see a black diamond. But in Notepad++ the first one is [NAK] and second one is [SYN]. For example escape character shows as [ESC] i.e. `\033`.

**Comment by mintty** [comment]
Oh, these are Ctrl+U and Ctrl+V which are internal single-character shortcuts for these options. As I included them verbatim in the source (could have used numeric notation instead, or a macro), it depends on your editor how they are displayed.

**Comment by mintty** [comment]
Released 1.9.0.
